patches/PSP.cmake

```diff
@@ -1,34 +1,64 @@
-SET(CMAKE_SYSTEM_NAME Generic)
-SET(CMAKE_SYSTEM_VERSION 1)
-SET(CMAKE_CROSSCOMPILING TRUE)
+#
+# CMake toolchain file for PSP.
+#
+# Copyright 2019 - Wally
+# Copyright 2020 - Daniel 'dbeef' Zalega
 
-# set compiler
-set(CMAKE_C_COMPILER psp-gcc)
-set(CMAKE_CXX_COMPILER psp-g++)
+# Determine PSP toolchain installation directory (psp-config binary is added to path after successful installation):
+execute_process(COMMAND bash -c "psp-config --pspdev-path" OUTPUT_VARIABLE PSPDEV OUTPUT_STRIP_TRAILING_WHITESPACE)
 
-SET(BUILD_SHARED_LIBS FALSE)
-SET(CMAKE_EXECUTABLE_SUFFIX ".elf")
+# Assert that PSP SDK path is now defined:
+if (NOT DEFINED PSPDEV)
+    message(FATAL_ERROR "PSPDEV not defined. Pass -DPSPDEV=/usr/local/pspdev or other installation location.")
+endif ()
 
-# set find root path
-execute_process(COMMAND psp-config --pspsdk-path
-  OUTPUT_VARIABLE PSPSDK_PATH
-  OUTPUT_STRIP_TRAILING_WHITESPACE)
+# Set helper variables:
+set(PSPSDK ${PSPDEV}/psp/sdk)
+set(PSPBIN ${PSPDEV}/bin)
+set(PSPCMAKE ${PSPDEV}/psp/share/cmake)
 
-execute_process(COMMAND psp-config --psp-prefix
-  OUTPUT_VARIABLE PSP_PREFIX
-  OUTPUT_STRIP_TRAILING_WHITESPACE)
+# Basic CMake Declarations - Required (CMAKE_SYSTEM_NAME sets cross compiling to true)
+set(CMAKE_SYSTEM_NAME Generic)
+set(TOOLCHAIN "${PSPDEV}")
+set(CMAKE_C_COMPILER ${CC})
+set(CMAKE_CXX_COMPILER ${CXX})
+set(CMAKE_FIND_ROOT_PATH "${PSPSDK};${PSPDEV}")
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
 
-set(CMAKE_FIND_ROOT_PATH "${PSPSDK_PATH};${PSP_PREFIX}")
-include_directories(SYSTEM "${PSPSDK_PATH}/include;${PSP_PREFIX}/include")
+# Include directories:
+include_directories(${include_directories} ${PSPDEV}/include ${PSPSDK}/include)
 
-SET(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
-SET(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
-SET(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+# Discard debug information:
+add_definitions("-G0")
+# Definitions that may be needed to use some libraries:
+add_definitions("-D__PSP__")
+add_definitions("-DHAVE_OPENGL")
 
-# reduce link error
-ADD_DEFINITIONS("-G0")
+# Set paths to toolchain binaries:
+set(CC ${PSPBIN}/psp-gcc)
+set(CXX ${PSPBIN}/psp-g++)
+set(AS ${PSPBIN}/psp-as)
+set(LD ${PSPBIN}/psp-gcc)
+set(AR ${PSPBIN}/psp-ar)
+set(RANLIB ${PSPBIN}/psp-ranlib)
+set(STRIP1 ${PSPBIN}/psp-strip)
+set(MKSFO ${PSPBIN}/mksfo)
+set(PACK_PBP ${PSPBIN}/pack-pbp)
+set(FIXUP ${PSPBIN}/psp-fixup-imports)
+set(ENC ${PSPBIN}/PrxEncrypter)
+
+# Pass these libraries to linker calls by default:
+set(PSP_LIBRARIES
```

Comment:
This library list seems suspicious:
- multiple calls to the c library
- having both user and kernel libraries mixed

Reply:
Will remove redundant C library link flags.

For the kernel/user libraries - it's either this, or leaving linking to specific libraries to the user.

Reply:
Removed redundant `-lc`.

What do you recommend about this list of libraries?

And besides, I don't know if you are aware of the PSP homebrew discord server, there is a channel specifically for psp-toolchain.
I am leaving the link, we would all be pleased to have you there (and would potentially improve communication).

https://discord.gg/MGs9NM

Reply:
I am available on discord now.

Problem with the c library is that we have multiple. Most users would want newlib, so this should be ok.
Either we use this default list or we need to have 3 or 4 lists that need alternation by some variable (which is ugly).